### ISSUE TITLE
The MorePackets check doesn't set players back properly

### ISSUE THREAD

**Issue by Lysandr0** [NoCheatPlus/Issues#432]
#### Short description of the issue.
On a continuous violation for MorePackets, the check sometimes does not cancel the player’s move, allowing a sort of Timer / Speed bypass. (Tested on both 1.8.8 and 1.11)

#### Complete output of the _ncp version_ command (please prefer text here).
---- Version information ----
#### Server ####
git-Spigot-21fe707-e1ebe52 MC: 1.8.8~
detected: 1.8.8
#### NoCheatPlus ####
 Plugin: 3.15.2-SNAPSHOT-sMD5NET-b1089
MCAccess: 1.8.4-1.8.8 / Spigot-CB-1.8_R3
Features:
blocks: BlocksMC1_4 | BlocksMC1_5 | BlocksMC1_6_1 | BlocksMC1_7_2 | BlocksMC1_8
checks: FastConsume | Gutenberg | AttackFrequency | FlyingFrequency | KeepAliveFrequency
 defaults: pvpKnockBackVelocity
 packet-listeners: UseEntityAdapter | MovingFlying | OutgoingPosition | KeepAliveAdapter | SoundDistance | CatchAllAdapter
Hooks: AllViolations NCP~ 1.0 | TestNCPHook 1.0
 #### Related Plugins ####
ProtocolLib v4.2.1-SNAPSHOT-b353

#### How to reproduce the issue (is it random / always happens / side conditions)?
 Just use a Timer hack (with a not too high speed) and walk around for a while.
It happens pretty much always.

#### Have you made any (noteworthy) changes to the default configuration of NCP?
No.

#### Do you run any other special plugins that affect game mechanics next to NC+ (adding skills, machines, adding/changing blocks, other anti cheating)?
No, tested on a sandbox server.

#### Provide additional information, such as error/debug/violation logs. (For reproducible issues with moving consider: https://github.com/NoCheatPlus/Docs/wiki/Debugging#on-the-fly-debug-output-for-individual-players) Use GitHub gists or http://pastebin.com/ or similar for huge logs, or attach a (zip-) file here. For image uploads, please prefer a site that allows viewing without JavaScript, you can upload file attachments for issues right here.
Here's a debug log.
[nocheatplus.txt](https://github.com/NoCheatPlus/Issues/files/967873/nocheatplus.txt)

**Comment by asofold** [comment]
We need to clarify / distinguish here:
1. For legacy versions of Minecraft (before 1.9 roughly) NCP will not cancel the moving event, but use another technique - just if the log gives rise to confusion. This could be configured, however the older versions behave differently and it would lead to increased freezing and collateral follow-up violations.
2. The best thing to do for comparison is the 'ncp stopwatch distance 100' command, for testing 100 blocks flat distance for instance (vary as needed). This way we can compare to what a vanilla client can do, enabling us to name some speed-up percentage amount. (Nice to have.)
3. Currently NCP will update the set-back every second or so. It does not always keep a one second old set-back (technically that would mean remembering 20 past moves at least) - so a timer cheat could try to abuse the set-back resetting. NCP will always use the oldest set-back, which could be survivalfly in case of jumping/in-air.
4. We need to clarify: Not setting back / not cancel coud mean putting the player on ground where they just had been one move before - in case the set back had just been updated, this would appear like not having set back (much).

So i have to somehow ensure if this is a bug (with how setting back is done), or expected behavior (just more or less refined abuse of the set-back updating). Best guess is the latter.

A mitigation attempt could be to inflate past move tracking to 20 moves (performance wise hardly a difference, a little more memory is used), which allows always using a pretty old location for morepackets set-back. The memory impact could also be reduced by only inflating the past moves storage temporarily, once the first violation of morepackets has happened + a cooldown. The latter is interesting (technically), because it's also planned to have this temporary inflation for other purposes, such as being able to log past moves in case of false positives, having a special-case trigger the inflation temporarily and another trigger will make the past moves be logged.

So i'll have to have another look at the log if i find actual bugs - and it would also be really helpful to have the stopwatch comparison for 50-100 blocks.

**Comment by Lysandr0** [comment]
So, i did some testing with the stopwatch command, here’s the results:
  Distance 100 with a Timer hack on (3.0 speed ~ A higher value would - much more frequently -trigger the FlyingFrequency check, best setting would be 2.5 more or less) + sprint:
￼![3bdf29a7edaca2d50186cc3cde8bf879](https://cloud.githubusercontent.com/assets/20213873/25583383/d3a38b40-2e91-11e7-9c19-ba82b83efe23.png)
 
 Distance 100 with a vanilla client (sprint + jump):
![9889bf236cf356bab0cbdfdc3f4f4edd](https://cloud.githubusercontent.com/assets/20213873/25583433/0c38d15e-2e92-11e7-99d9-edc9c949310c.png)

NCP version:
![9e87330dc72e4d55306f2bb57dad381b](https://cloud.githubusercontent.com/assets/20213873/25584026/aa9b6df0-2e94-11e7-9283-0e3b5162422a.png)



(I removed the kick reached at 100 VL on MorePackets to better reproduce the issue)


You can roughly see how it looks like with this gif:  https://gyazo.com/34ce3fa767a3f4fb0030652df0b8c867

**Comment by asofold** [comment]
That would be too much speed up to accept :). All in all this looks like the set back is not done right.

I have to ask what i lost track of last message:
* Do you have morepackets or survivalfly set so it doesn't cancel for low violation levels?
* Has morepacket limits been changed?
* FlyingFrequency does not seem to be enabled in that 'ncp version' output - is that right?

Ok i'll also see from the log once i really dig into it (most likely)...

----

Edit: Clearly the set-back updating is problematic somehow, accoridng to the log.

**Comment by asofold** [comment]
Build 1091 contains a fix attempt (reintroduce a set back age for morepackets, configurable with a default of 40 moves, i.e. 2 thinkable seconds).

**Comment by Lysandr0** [comment]
Works great again with build #1091 :)
 i will close the ticket for now. (if i find something else i' ll leave a note here)

**Comment by asofold** [comment]
Thank you for reporting :).
